test/core/security/grpc_tls_certificate_provider_test.cc

```diff
@@ -495,6 +495,67 @@ TEST_F(GrpcTlsCertificateProviderTest,
   CancelWatch(watcher_state_1);
 }
 
+TEST_F(GrpcTlsCertificateProviderTest,
+       FileWatcherCertificateProviderWithInvalidCertificateKeyPairOnRefresh) {
+  // Create temporary files and copy cert data into them.
+  TmpFile tmp_root_cert(root_cert_);
+  TmpFile tmp_identity_key(private_key_);
+  TmpFile tmp_identity_cert(cert_chain_);
+  // Create FileWatcherCertificateProvider.
+  FileWatcherCertificateProvider provider(tmp_identity_key.name(),
+                                          tmp_identity_cert.name(),
+                                          tmp_root_cert.name(), 1);
+  WatcherState* watcher_state_1 =
+      MakeWatcher(provider.distributor(), kCertName, kCertName);
+  // Expect to see the credential data.
+  EXPECT_THAT(watcher_state_1->GetCredentialQueue(),
+              ::testing::ElementsAre(CredentialInfo(
+                  root_cert_, MakeCertKeyPairs(private_key_.c_str(),
+                                               cert_chain_.c_str()))));
+  // Copy new data to files.
+  // TODO(ZhenLian): right now it is not completely atomic. Use the real atomic
+  // update when the directory renaming is added in gpr.
```

Comment:
@markdroth Can we delete the TODO here? Because I think now we don't need the atomic update. Early on we need this because there could be the time when the credentials are updated half way in the middle while our refreshing thread is reading.
Now since we have the key-cert mismatch check, even if that happens, the current content will be disregarded, and picked up again next time(as the test logic says).

Reply:
should I go ahead with the changes and delete the TODO?

Reply:
@ZhenLian

Reply:
Let's keep it for now - it might change later though